File: internal/skills/skills.go
// Package skills loads the project-local reverse engineering workflows from
// skills/<name>/SKILL.md (falling back to the copies embedded in the binary).
package skills

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"

	"github.com/overkazaf/re-agent/internal/assets"
	"github.com/overkazaf/re-agent/internal/util"
)

type Skill struct {
	Name        string
	Description string
	Path        string
	Body        string
	Tags        []string
}

// Load reads every skill directory. Missing or unreadable skills are skipped
// rather than fatal.
func Load() []Skill {
	return loadFrom(assets.SkillsDir(), assets.EmbeddedSkills())
}

func loadFrom(dir string, embedded map[string]string) []Skill {
	byName := map[string]Skill{}
	for name, body := range embedded {
		skill := parse(name, "embedded:skills/"+name+"/SKILL.md", body)
		byName[strings.ToLower(skill.Name)] = skill
	}
	if dir != "" {
		entries, err := os.ReadDir(dir)
		if err == nil {
			for _, entry := range entries {
				if !entry.IsDir() {
					continue
				}
				skillPath := filepath.Join(dir, entry.Name(), "SKILL.md")
				data, err := os.ReadFile(skillPath)
				if err != nil {
					continue
				}
				skill := parse(entry.Name(), skillPath, string(data))
				byName[strings.ToLower(skill.Name)] = skill
			}
		}
	}
	out := make([]Skill, 0, len(byName))
	for _, skill := range byName {
		out = append(out, skill)
	}
	sort.Slice(out, func(a, b int) bool { return out[a].Name < out[b].Name })
	return out
}

func parse(dirName, path, body string) Skill {
	metadata := parseFrontmatter(body)
	name := metadata["name"]
	if name == "" {
		name = dirName
	}
	description := metadata["description"]
	if description == "" {
		description = firstHeading(body)
	}
	if description == "" {
		description = "Built-in reverse engineering workflow."
	}
	tagSource := metadata["tags"]
	if tagSource == "" {
		tagSource = dirName
	}
	return Skill{Name: name, Description: description, Path: path, Body: body, Tags: splitTags(tagSource)}
}

func Find(list []Skill, name string) *Skill {
	needle := strings.ToLower(name)
	for i := range list {
		if strings.ToLower(list[i].Name) == needle {
			return &list[i]
		}
		for _, tag := range list[i].Tags {
			if strings.ToLower(tag) == needle {
				return &list[i]
			}
		}
	}
	return nil
}

func FormatList(list []Skill) string {
	if len(list) == 0 {
		return "No built-in skills found."
	}
	rows := make([]string, 0, len(list))
	for _, skill := range list {
		rows = append(rows, fmt.Sprintf("%-18s %s", skill.Name, skill.Description))
	}
	return strings.Join(rows, "\n")
}

// SystemPrompt is the catalog appended to the agent's system prompt.
func SystemPrompt(list []Skill) string {
	if len(list) == 0 {
		return ""
	}
	catalog := make([]string, 0, len(list))
	for _, skill := range list {
		tags := ""
		if len(skill.Tags) > 0 {
			tags = " Tags: " + strings.Join(skill.Tags, ", ")
		}
		catalog = append(catalog, fmt.Sprintf("- %s: %s%s", skill.Name, skill.Description, tags))
	}
	return strings.Join([]string{
		"",
		"## Built-in 0xAF-Re Skills",
		"",
		"The host has project-local reverse engineering skills. Use them when a task matches their scope.",
		"Ask for `read_skill` when you need full instructions; use `list_skills` to inspect the catalog.",
		"Operators can run `/skills` to list them or `/skill <name> <task>` to force one workflow for a turn.",
		"",
		strings.Join(catalog, "\n"),
	}, "\n")
}

// TurnPrompt forces one skill's workflow for a single turn.
func TurnPrompt(skill Skill, task string) string {
	return strings.Join([]string{
		"Use built-in skill: " + skill.Name,
		"",
		util.Clip(skill.Body, turnPromptSkillLimit),
		"",
		"Task:",
		strings.TrimSpace(task),
	}, "\n")
}

const turnPromptSkillLimit = 32_000

var (
	frontmatterRE = regexp.MustCompile(`^([A-Za-z0-9_-]+):\s*(.*)$`)
	headingRE     = regexp.MustCompile(`(?m)^#\s+(.+)$`)
	tagSplitRE    = regexp.MustCompile(`[,\s]+`)
)

func parseFrontmatter(body string) map[string]string {
	out := map[string]string{}
	if !strings.HasPrefix(body, "---\n") {
		return out
	}
	end := strings.Index(body[4:], "\n---")
	if end < 0 {
		return out
	}
	for _, line := range util.Lines(body[4 : 4+end]) {
		match := frontmatterRE.FindStringSubmatch(line)
		if match == nil {
			continue
		}
		out[match[1]] = strings.TrimSpace(strings.Trim(match[2], `"'`))
	}
	return out
}

func firstHeading(body string) string {
	match := headingRE.FindStringSubmatch(body)
	if match == nil {
		return ""
	}
	return strings.TrimSpace(match[1])
}

func splitTags(value string) []string {
	var out []string
	for _, tag := range tagSplitRE.Split(value, -1) {
		if trimmed := strings.TrimSpace(tag); trimmed != "" {
			out = append(out, trimmed)
		}
	}
	return out
}

File: src/skills.ts
import * as fs from "node:fs/promises";
import * as path from "node:path";
import { fileURLToPath } from "node:url";
import { clip } from "./utils";

export interface BuiltInSkill {
  name: string;
  description: string;
  path: string;
  body: string;
  tags: string[];
}

export function projectRoot(): string {
  return path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..");
}

export function builtInSkillsRoot(): string {
  return path.join(projectRoot(), "skills");
}

export async function loadBuiltInSkills(root = builtInSkillsRoot()): Promise<BuiltInSkill[]> {
  const entries = await fs.readdir(root, { withFileTypes: true }).catch(() => []);
  const skills: BuiltInSkill[] = [];
  for (const entry of entries) {
    if (!entry.isDirectory()) continue;
    const skillPath = path.join(root, entry.name, "SKILL.md");
    const body = await fs.readFile(skillPath, "utf8").catch(() => undefined);
    if (!body) continue;
    const metadata = parseFrontmatter(body);
    skills.push({
      name: metadata.name || entry.name,
      description: metadata.description || firstHeading(body) || "Built-in reverse engineering workflow.",
      tags: splitTags(metadata.tags || entry.name),
      path: skillPath,
      body,
    });
  }
  return skills.sort((a, b) => a.name.localeCompare(b.name));
}

export function findBuiltInSkill(skills: BuiltInSkill[], name: string): BuiltInSkill | undefined {
  const needle = name.toLowerCase();
  return skills.find(skill => skill.name.toLowerCase() === needle || skill.tags.some(tag => tag.toLowerCase() === needle));
}

export function formatSkillList(skills: BuiltInSkill[]): string {
  if (skills.length === 0) return "No built-in skills found.";
  return skills.map(skill => `${skill.name.padEnd(18)} ${skill.description}`).join("\n");
}

export function skillSystemPrompt(skills: BuiltInSkill[]): string {
  if (skills.length === 0) return "";
  const catalog = skills.map(skill => `- ${skill.name}: ${skill.description}`).join("\n");
  return [
    "",
    "## Built-in 0xAF-Re Skills",
    "",
    "The host has project-local reverse engineering skills. Use them when a task matches their scope.",
    "Ask for `read_skill` when you need full instructions; use `list_skills` to inspect the catalog.",
    "Operators can run `/skills` to list them or `/skill <name> <task>` to force one workflow for a turn.",
    "",
    catalog,
  ].join("\n");
}

export function skillTurnPrompt(skill: BuiltInSkill, task: string): string {
  return [
    `Use built-in skill: ${skill.name}`,
    "",
    clip(skill.body, 20_000),
    "",
    "Task:",
    task.trim(),
  ].join("\n");
}

function parseFrontmatter(body: string): Record<string, string> {
  if (!body.startsWith("---\n")) return {};
  const end = body.indexOf("\n---", 4);
  if (end < 0) return {};
  const out: Record<string, string> = {};
  for (const line of body.slice(4, end).split(/\r?\n/)) {
    const match = /^([A-Za-z0-9_-]+):\s*(.*)$/.exec(line);
    if (match) out[match[1]] = match[2].replace(/^["']|["']$/g, "").trim();
  }
  return out;
}

function firstHeading(body: string): string | undefined {
  const match = /^#\s+(.+)$/m.exec(body);
  return match?.[1]?.trim();
}

function splitTags(value: string): string[] {
  return value.split(/[,\s]+/).map(tag => tag.trim()).filter(Boolean);
}
